utils/writeable/src/utf.rs

```diff
@@ -0,0 +1,89 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{impl_display_with_writeable, LengthHint, Writeable};
+
+use core::fmt;
+
+/// Implements [`Writeable`] for [`&[u8]`] according to the [WHATWG Encoding Standard](
+/// https://encoding.spec.whatwg.org/#utf-8-decoder).
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)] // newtype
+pub struct PotentiallyInvalidUtf8<'a>(pub &'a [u8]);
+
+impl Writeable for PotentiallyInvalidUtf8<'_> {
+    fn write_to<W: core::fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        let mut remaining = self.0;
+        while !remaining.is_empty() {
+            match core::str::from_utf8(remaining) {
+                Ok(str) => {
+                    return sink.write_str(str);
+                }
+                Err(e) => {
+                    let (str, rest) = remaining.split_at(e.valid_up_to());
+                    sink.write_str(unsafe { core::str::from_utf8_unchecked(str) })?;
+                    sink.write_char(char::REPLACEMENT_CHARACTER)?;
+                    match e.error_len() {
+                        None => remaining = &[],
+                        #[allow(clippy::indexing_slicing)] // l guaranteed to be in rest
+                        Some(l) => remaining = &rest[l..],
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn writeable_length_hint(&self) -> crate::LengthHint {
+        // Lower bound is all ASCII, upper bound is all replacement character
```

Comment:
This is good if we want `crate::LengthHint` to be strict, but if we want it to be an approximate upper bound, it would be more useful for this to be perhaps only a little bit higher than the success case. `*3` can be quite large and is almost never necessary.

Also:

```suggestion
        // Lower bound is all valid UTF-8, upper bound is all replacement character
```